GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,208 @@
+            {
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCert, out GitConfigSetting sslCerts))
+                {
+                    this.certificatePathOrSubjectCommonName = sslCerts.Values.Single();
+                }
+
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCertPasswordProtected, out GitConfigSetting isSslCertPasswordProtected))
+                {
+                    this.isCertificatePasswordProtected = isSslCertPasswordProtected.Values.Select(bool.Parse).Single();
+                }
+
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslVerify, out GitConfigSetting sslVerify))
+                {
+                    this.ShouldVerify = sslVerify.Values.Select(bool.Parse).Single();
+                }
+            }
+        }
+
+        public bool ShouldVerify { get; }
+
+        public string GetCertificatePassword(ITracer tracer, GitProcess git)
+        {
+            if (git.TryGetCertificatePassword(tracer, this.certificatePathOrSubjectCommonName, out string password, out string error))
+            {
+                return password;
+            }
+
+            return null;
+        }
+
+        public X509Certificate2 GetCertificate(ITracer tracer, GitProcess gitProcess)
+        {
+            if (string.IsNullOrEmpty(this.certificatePathOrSubjectCommonName))
+            {
+                return null;
+            }
+
+            EventMetadata metadata = new EventMetadata
+            {
+                { "CertificatePathOrSubjectCommonName", this.certificatePathOrSubjectCommonName },
+                { "IsCertificatePasswordProtected", this.isCertificatePasswordProtected },
+                { "ShouldVerify", this.ShouldVerify }
+            };
+
+            X509Certificate2 result =
+                this.GetCertificateFromFile(tracer, metadata, gitProcess) ??
+                this.GetCertificateFromStore(tracer, metadata);
+
+            if (result == null)
+            {
+                tracer.RelatedError("Certificate {0} not found", this.certificatePathOrSubjectCommonName);
+            }
+
+            return result;
+        }
+
+        private static void LogWithAppropriateLevel(ITracer tracer, EventMetadata metadata, IEnumerable<X509Certificate2> certificates, string logMessage)
+        {
+            Action<EventMetadata, string> loggingFunction;
+            int numberOfCertificates = certificates.Count();
+
+            switch (numberOfCertificates)
+            {
+                case 0:
+                    loggingFunction = tracer.RelatedError;
+                    break;
+                case 1:
+                    loggingFunction = tracer.RelatedInfo;
+                    break;
+                default:
+                    loggingFunction = tracer.RelatedWarning;
+                    break;
+            }
+
+            loggingFunction(
+                metadata,
+                logMessage);

```

Comment:
We can get rid of the `loggingFunction` now and just call the `tracer.Related` method in each of the case statements.